### ISSUE TITLE
WebSocketSubject-based component: Identifier 'Context' has already been declared

### ISSUE THREAD

**Issue by kael** [ionic-team/stencil#505]
**Stencil version:**

```
 @stencil/core@0.3.0-5
 @ionic/core@0.0.2-48
 rxjs@5.5.6
 NPM 5.6.0
 Node v8.9.4
 OS Ubuntu 16.04.3 LTS
```

**I'm submitting a:**

- [x] bug report
- [ ] feature request
- [ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or https://stencil-worldwide.slack.com

**Current behavior:**

When using a Websocket client component based on [WebSocketSubject](http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#static-method-webSocket), the following error is raised in the browser:

`app.core.js:2068 SyntaxError: Identifier 'Context' has already been declared`

Which points to this part of the code:

```js
    function loadBundle(cmpMeta, modeName, cb) {
      if (cmpMeta.componentConstructor) {
        // we're already all loaded up :)
        cb();
      } else {
        const bundleId = 'string' === typeof cmpMeta.bundleIds ? cmpMeta.bundleIds : cmpMeta.bundleIds[modeName];
        const url = publicPath + bundleId + (useScopedCss(domApi.$supportsShadowDom, cmpMeta) ? '.sc' : '') + '.js';
        // dynamic es module import() => woot!
                import(url).then(importedModule => {
          try {
            // async loading of the module is done
            cmpMeta.componentConstructor || (
            // we haven't initialized the component module yet
            // get the component constructor from the module
            cmpMeta.componentConstructor = importedModule[dashToPascalCase(cmpMeta.tagNameMeta)]);
            // initialize this components styles
            // it is possible for the same component to have difficult styles applied in the same app
                        initStyleTemplate(domApi, cmpMeta.componentConstructor);
          } catch (e) {
            // oh man, something's up
            console.error(e);
            // provide a bogus component constructor
            // so the rest of the app acts as normal
                        cmpMeta.componentConstructor = class {};
          }
          // bundle all loaded up, let's continue
                    cb();
        }).catch(err => console.error(err)); // <= Line 2068
      }
    }
    true;
    plt.attachStyles = attachStyles;
    return plt;
  }
  /*
```

**Expected behavior:**

**Steps to reproduce:**

**Related code:**

The component is an extended class based on service-like  (à la Ionic or Angular) WS client (the WS client code is running fine in an Ionic 3 PWA):

* stencil.config.js

```sx
exports.config = {
  bundles: [
    { components: ['my-app', 'app-home'] },
    { components: ['app-profile'] },
    { components: ['ws-client'] },
  ],
  collections: [
    { name: '@stencil/router' },
    { name: '@ionic/core' },
  ],
  serviceWorker: {
    swSrc: 'src/sw.js'
  },
  globalStyle: 'src/global/app.css'
};

exports.devServer = {
  root: 'www',
  watchGlob: '**/**'
};
```

* src/providers/ws-service.tsx

```tsx
import { Observable } from 'rxjs/Rx';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import { WebSocketSubject } from 'rxjs/observable/dom/WebSocketSubject';

export namespace NSIWebsocketService {
  export type status = 'disconnected'|'connecting'|'connected';
}

export interface IWebsocketService {
  connect: () => any;
}

export class WebsocketService implements IWebsocketService {

  protected connection:WebSocketSubject<any>;
  protected maxRetries:number = 1000;
  public connected:boolean = false;
  public reconnected:boolean = false;
  protected _messageId:number = 0;
  protected get messageId():number {
    this._messageId++;
    return this._messageId;
  }
  protected get uri():string { return `wss://${window.location.hostname}/ws`; }

  public status:NSIWebsocketService.status = 'disconnected';

  private _open$:BehaviorSubject<any>;
  public get open$() { return this._open$.asObservable().share(); }
  public get capability() { return 'WebSocket' in window; }

  constructor() {
    this._open$ = new BehaviorSubject({type:false});
  }

  connect() {
    console.log('WS connect');
    this.status = 'connecting';
    this.connection = Observable.webSocket({
      url: this.uri,
      openObserver: this._open$
    });
    this.open$.subscribe(event => {
      console.log('[WS] Open Event', event);
      if (event.type === 'connected') this.connected = true;
      this.status = event.type ? 'connected' : 'disconnected';
    });
  }

}
```

* src/components/ws-client/ws-client.tsx

```tsx
import {
  Component,
  Event,
  EventEmitter,
  Method,
  Prop,
  State,
  Watch
} from '@stencil/core';

import {
  NSIWebsocketService,
  IWebsocketService,
  WebsocketService
} from '../../providers/ws-service';

@Component({
  tag: 'ws-client',
  // styleUrl: 'ws-client.scss'
})
export class WSClient extends WebsocketService implements IWebsocketService {

  constructor() {
    super();
  }

  componentWillLoad() {
    console.log('The component is about to be rendered');
    if (this.capability)  this.connect();
  }

  render() {
    return null;
  }

}
```

* src/components/app-home/app-home.tsx

```tsx
import { Component } from '@stencil/core';


@Component({
  tag: 'app-home',
  styleUrl: 'app-home.scss'
})
export class AppHome {

  render() {
    return (
      <ion-page class='show-page'>
        <ion-header md-height='56px'>
          <ion-toolbar color='primary'>
            <ion-title>Ionic PWA Toolkit</ion-title>
          </ion-toolbar>
        </ion-header>
        <ion-content>
          <ws-client></ws-client>
        </ion-content>
      </ion-page>
    );
  }
}

```

**Other information:**

**Comment by jgw96** [comment]
Hmmm yeah, im guessing this is caused by us having a `context` variable we use across Stencil. @adamdbradley not sure if we want to rename that var?

**Comment by adamdbradley** [comment]
Yeah and I think we can remove `Context` entirely now with es module support. @jthoms1 see any issues with that for stencil router and redux?

**Comment by kael** [comment]
Thanks for looking at it.

Is there a best practice to avoid such collisions ? Can we use TS namespaces somehow ? Or prefixing ? Looks like a generic issue though.

**Comment by kael** [comment]
BTW, as an aside, I'm not sure using `WebSocketSubject` is still relevant considering we can use `@Watch` and `@Status` (not to say that it shouldn't be fixed).

Wondering if Ionic or Stencil could include some built-in elements like `<ionic-http/>` or `<ionic-ws/>`. 

Also, on the matter of Ionic/Angular providers, would include, in the

**Comment by jthoms1** [comment]
This issue should be resolved with commit fe66a3faa819c2f2850c4d674dc4280345bb64b6.
